NPE when ServiceKnot has not defined params

When I try to override `ServiceKnot` configuration in my project:
```json
  "config": {
    "knotx:io.knotx.ServiceKnot": {
      "options": {
        "config": {
          "services": [
            {
              "name": "booksdata",
              "address": "knotx.adapter.service.custom"
            }
          ]
        }
      }
    }
  }
```
I get NPE unless `params` property is not defined in a service:
```
2017-03-08 08:38:17 [vert.x-eventloop-thread-5] ERROR i.k.k.s.impl.ServiceKnotProxyImpl - Error happened during Template processing
java.lang.NullPointerException: null
        at io.knotx.knot.service.service.ServiceEntry.mergeParams(ServiceEntry.java:49) ~[knotx-standalone-1.0.0.fat.jar:na]
        at io.knotx.knot.service.service.ServiceEngine.lambda$mergeWithConfiguration$1(ServiceEngine.java:68) ~[knotx-standalone-1.0.0.fat.jar:na]
        at java.util.Optional.map(Unknown Source) ~[na:1.8.0_111]
        at io.knotx.knot.service.service.ServiceEngine.mergeWithConfiguration(ServiceEngine.java:64) ~[knotx-standalone-1.0.0.fat.jar:na]
        at rx.internal.operators.OnSubscribeMap$MapSubscriber.onNext(OnSubscribeMap.java:69) [custom-service-adapter-1.0.0-fat.jar:na]
        at rx.observers.Subscribers$5.onNext(Subscribers.java:235) ~[custom-service-adapter-1.0.0-fat.jar:na]
        at rx.internal.operators.OnSubscribeFromIterable$IterableProducer.fastpath(OnSubscribeFromIterable.java:173) ~[custom-service-adapter-1.0.0-fat.jar:na]
        at rx.internal.operators.OnSubscribeFromIterable$IterableProducer.request(OnSubscribeFromIterable.java:86) ~[custom-service-adapter-1.0.0-fat.jar:na]
        at rx.Subscriber.setProducer(Subscriber.java:211) ~[custom-service-adapter-1.0.0-fat.jar:na]
        at rx.internal.operators.OnSubscribeMap$MapSubscriber.setProducer(OnSubscribeMap.java:102) [custom-service-adapter-1.0.0-fat.jar:na]
        at rx.Subscriber.setProducer(Subscriber.java:205) ~[custom-service-adapter-1.0.0-fat.jar:na]
        at rx.internal.operators.OnSubscribeMap$MapSubscriber.setProducer(OnSubscribeMap.java:102) [custom-service-adapter-1.0.0-fat.jar:na]
        at rx.Subscriber.setProducer(Subscriber.java:205) ~[custom-service-adapter-1.0.0-fat.jar:na]
        at rx.internal.operators.OnSubscribeFromIterable.call(OnSubscribeFromIterable.java:63) ~[custom-service-adapter-1.0.0-fat.jar:na]
        at rx.internal.operators.OnSubscribeFromIterable.call(OnSubscribeFromIterable.java:34) ~[custom-service-adapter-1.0.0-fat.jar:na]
        at rx.Observable.unsafeSubscribe(Observable.java:10150) ~[custom-service-adapter-1.0.0-fat.jar:na]
        at rx.internal.util.ScalarSynchronousObservable$3.call(ScalarSynchronousObservable.java:235) ~[custom-service-adapter-1.0.0-fat.jar:na]
        at rx.internal.util.ScalarSynchronousObservable$3.call(ScalarSynchronousObservable.java:228) ~[custom-service-adapter-1.0.0-fat.jar:na]
        at rx.Observable.unsafeSubscribe(Observable.java:10150) ~[custom-service-adapter-1.0.0-fat.jar:na]
        at rx.internal.operators.OnSubscribeMap.call(OnSubscribeMap.java:48) ~[custom-service-adapter-1.0.0-fat.jar:na]
        at rx.internal.operators.OnSubscribeMap.call(OnSubscribeMap.java:33) ~[custom-service-adapter-1.0.0-fat.jar:na]
        at rx.Observable.unsafeSubscribe(Observable.java:10150) ~[custom-service-adapter-1.0.0-fat.jar:na]
        at rx.internal.operators.OnSubscribeDoOnEach.call(OnSubscribeDoOnEach.java:41) ~[custom-service-adapter-1.0.0-fat.jar:na]
        at rx.internal.operators.OnSubscribeDoOnEach.call(OnSubscribeDoOnEach.java:30) ~[custom-service-adapter-1.0.0-fat.jar:na]
        at rx.Observable.unsafeSubscribe(Observable.java:10150) ~[custom-service-adapter-1.0.0-fat.jar:na]
        at rx.internal.operators.OnSubscribeMap.call(OnSubscribeMap.java:48) ~[custom-service-adapter-1.0.0-fat.jar:na]
        at rx.internal.operators.OnSubscribeMap.call(OnSubscribeMap.java:33) ~[custom-service-adapter-1.0.0-fat.jar:na]
        at rx.internal.operators.OnSubscribeLift.call(OnSubscribeLift.java:48) ~[custom-service-adapter-1.0.0-fat.jar:na]
        at rx.internal.operators.OnSubscribeLift.call(OnSubscribeLift.java:30) ~[custom-service-adapter-1.0.0-fat.jar:na]
        at rx.Observable.unsafeSubscribe(Observable.java:10150) ~[custom-service-adapter-1.0.0-fat.jar:na]
        at rx.internal.operators.DeferredScalarSubscriber.subscribeTo(DeferredScalarSubscriber.java:153) ~[custom-service-adapter-1.0.0-fat.jar:na]
        at rx.internal.operators.OnSubscribeReduceSeed.call(OnSubscribeReduceSeed.java:40) ~[custom-service-adapter-1.0.0-fat.jar:na]
        at rx.internal.operators.OnSubscribeReduceSeed.call(OnSubscribeReduceSeed.java:24) ~[custom-service-adapter-1.0.0-fat.jar:na]
        at rx.Observable.unsafeSubscribe(Observable.java:10150) ~[custom-service-adapter-1.0.0-fat.jar:na]
        at rx.internal.operators.OnSubscribeMap.call(OnSubscribeMap.java:48) ~[custom-service-adapter-1.0.0-fat.jar:na]
        at rx.internal.operators.OnSubscribeMap.call(OnSubscribeMap.java:33) ~[custom-service-adapter-1.0.0-fat.jar:na]
        at rx.Observable.unsafeSubscribe(Observable.java:10150) ~[custom-service-adapter-1.0.0-fat.jar:na]
        at rx.internal.operators.OperatorMerge$MergeSubscriber.onNext(OperatorMerge.java:250) ~[custom-service-adapter-1.0.0-fat.jar:na]
        at rx.internal.operators.OperatorMerge$MergeSubscriber.onNext(OperatorMerge.java:147) ~[custom-service-adapter-1.0.0-fat.jar:na]
        at rx.internal.operators.OnSubscribeMap$MapSubscriber.onNext(OnSubscribeMap.java:77) [custom-service-adapter-1.0.0-fat.jar:na]
        at rx.internal.operators.OperatorMerge$MergeSubscriber.emitScalar(OperatorMerge.java:399) ~[custom-service-adapter-1.0.0-fat.jar:na]
        at rx.internal.operators.OperatorMerge$MergeSubscriber.tryEmit(OperatorMerge.java:357) ~[custom-service-adapter-1.0.0-fat.jar:na]
        at rx.internal.operators.OperatorMerge$InnerSubscriber.onNext(OperatorMerge.java:852) ~[custom-service-adapter-1.0.0-fat.jar:na]
        at io.knotx.knot.service.impl.ServiceKnotProxyImpl.lambda$compileHtmlFragment$5(ServiceKnotProxyImpl.java:88) ~[knotx-standalone-1.0.0.fat.jar:na]
        at rx.Observable.unsafeSubscribe(Observable.java:10150) ~[custom-service-adapter-1.0.0-fat.jar:na]
        at rx.internal.operators.OperatorMerge$MergeSubscriber.onNext(OperatorMerge.java:250) ~[custom-service-adapter-1.0.0-fat.jar:na]
        at rx.internal.operators.OperatorMerge$MergeSubscriber.onNext(OperatorMerge.java:147) ~[custom-service-adapter-1.0.0-fat.jar:na]
        at rx.internal.operators.OnSubscribeMap$MapSubscriber.onNext(OnSubscribeMap.java:77) [custom-service-adapter-1.0.0-fat.jar:na]
        at rx.internal.operators.OnSubscribeDoOnEach$DoOnEachSubscriber.onNext(OnSubscribeDoOnEach.java:101) ~[custom-service-adapter-1.0.0-fat.jar:na]
        at rx.internal.operators.OnSubscribeFilter$FilterSubscriber.onNext(OnSubscribeFilter.java:76) ~[custom-service-adapter-1.0.0-fat.jar:na]
        at rx.internal.operators.OnSubscribeFromIterable$IterableProducer.fastpath(OnSubscribeFromIterable.java:173) ~[custom-service-adapter-1.0.0-fat.jar:na]
        at rx.internal.operators.OnSubscribeFromIterable$IterableProducer.request(OnSubscribeFromIterable.java:86) ~[custom-service-adapter-1.0.0-fat.jar:na]
        at rx.Subscriber.setProducer(Subscriber.java:211) ~[custom-service-adapter-1.0.0-fat.jar:na]
        at rx.internal.operators.OnSubscribeMap$MapSubscriber.setProducer(OnSubscribeMap.java:102) [custom-service-adapter-1.0.0-fat.jar:na]
        at rx.Subscriber.setProducer(Subscriber.java:205) ~[custom-service-adapter-1.0.0-fat.jar:na]
        at rx.internal.operators.OnSubscribeFilter$FilterSubscriber.setProducer(OnSubscribeFilter.java:104) ~[custom-service-adapter-1.0.0-fat.jar:na]
        at rx.internal.operators.OnSubscribeFromIterable.call(OnSubscribeFromIterable.java:63) ~[custom-service-adapter-1.0.0-fat.jar:na]
        at rx.internal.operators.OnSubscribeFromIterable.call(OnSubscribeFromIterable.java:34) ~[custom-service-adapter-1.0.0-fat.jar:na]
        at rx.Observable.unsafeSubscribe(Observable.java:10150) ~[custom-service-adapter-1.0.0-fat.jar:na]
        at rx.internal.operators.OnSubscribeFilter.call(OnSubscribeFilter.java:45) ~[custom-service-adapter-1.0.0-fat.jar:na]
        at rx.internal.operators.OnSubscribeFilter.call(OnSubscribeFilter.java:30) ~[custom-service-adapter-1.0.0-fat.jar:na]
        at rx.Observable.unsafeSubscribe(Observable.java:10150) ~[custom-service-adapter-1.0.0-fat.jar:na]
        at rx.internal.operators.OnSubscribeDoOnEach.call(OnSubscribeDoOnEach.java:41) ~[custom-service-adapter-1.0.0-fat.jar:na]
        at rx.internal.operators.OnSubscribeDoOnEach.call(OnSubscribeDoOnEach.java:30) ~[custom-service-adapter-1.0.0-fat.jar:na]
        at rx.Observable.unsafeSubscribe(Observable.java:10150) ~[custom-service-adapter-1.0.0-fat.jar:na]
        at rx.internal.operators.OnSubscribeMap.call(OnSubscribeMap.java:48) ~[custom-service-adapter-1.0.0-fat.jar:na]
        at rx.internal.operators.OnSubscribeMap.call(OnSubscribeMap.java:33) ~[custom-service-adapter-1.0.0-fat.jar:na]
        at rx.internal.operators.OnSubscribeLift.call(OnSubscribeLift.java:48) ~[custom-service-adapter-1.0.0-fat.jar:na]
        at rx.internal.operators.OnSubscribeLift.call(OnSubscribeLift.java:30) ~[custom-service-adapter-1.0.0-fat.jar:na]
        at rx.Observable.unsafeSubscribe(Observable.java:10150) ~[custom-service-adapter-1.0.0-fat.jar:na]
        at rx.internal.operators.OnSubscribeMap.call(OnSubscribeMap.java:48) ~[custom-service-adapter-1.0.0-fat.jar:na]
        at rx.internal.operators.OnSubscribeMap.call(OnSubscribeMap.java:33) ~[custom-service-adapter-1.0.0-fat.jar:na]
        at rx.internal.operators.OnSubscribeLift.call(OnSubscribeLift.java:48) ~[custom-service-adapter-1.0.0-fat.jar:na]
        at rx.internal.operators.OnSubscribeLift.call(OnSubscribeLift.java:30) ~[custom-service-adapter-1.0.0-fat.jar:na]
        at rx.internal.operators.OnSubscribeLift.call(OnSubscribeLift.java:48) ~[custom-service-adapter-1.0.0-fat.jar:na]
        at rx.internal.operators.OnSubscribeLift.call(OnSubscribeLift.java:30) ~[custom-service-adapter-1.0.0-fat.jar:na]
        at rx.Observable.unsafeSubscribe(Observable.java:10150) ~[custom-service-adapter-1.0.0-fat.jar:na]
        at rx.internal.operators.OnSubscribeMap.call(OnSubscribeMap.java:48) ~[custom-service-adapter-1.0.0-fat.jar:na]
        at rx.internal.operators.OnSubscribeMap.call(OnSubscribeMap.java:33) ~[custom-service-adapter-1.0.0-fat.jar:na]
        at rx.internal.operators.OnSubscribeLift.call(OnSubscribeLift.java:48) ~[custom-service-adapter-1.0.0-fat.jar:na]
        at rx.internal.operators.OnSubscribeLift.call(OnSubscribeLift.java:30) ~[custom-service-adapter-1.0.0-fat.jar:na]
        at rx.Observable.subscribe(Observable.java:10246) ~[custom-service-adapter-1.0.0-fat.jar:na]
        at rx.Observable.subscribe(Observable.java:10213) ~[custom-service-adapter-1.0.0-fat.jar:na]
        at rx.Observable.subscribe(Observable.java:10053) ~[custom-service-adapter-1.0.0-fat.jar:na]
        at io.knotx.knot.AbstractKnotProxy.process(AbstractKnotProxy.java:47) ~[custom-service-adapter-1.0.0-fat.jar:na]
        at io.knotx.proxy.KnotProxyVertxProxyHandler.handle(KnotProxyVertxProxyHandler.java:127) ~[custom-service-adapter-1.0.0-fat.jar:na]
        at io.knotx.proxy.KnotProxyVertxProxyHandler.handle(KnotProxyVertxProxyHandler.java:52) ~[custom-service-adapter-1.0.0-fat.jar:na]
        at io.vertx.core.eventbus.impl.HandlerRegistration.deliver(HandlerRegistration.java:212) ~[custom-service-adapter-1.0.0-fat.jar:na]
        at io.vertx.core.eventbus.impl.HandlerRegistration.handle(HandlerRegistration.java:191) ~[custom-service-adapter-1.0.0-fat.jar:na]
        at io.vertx.core.eventbus.impl.EventBusImpl.lambda$deliverToHandler$3(EventBusImpl.java:503) ~[custom-service-adapter-1.0.0-fat.jar:na]
        at io.vertx.core.impl.ContextImpl.lambda$wrapTask$2(ContextImpl.java:316) ~[custom-service-adapter-1.0.0-fat.jar:na]
        at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:163) ~[custom-service-adapter-1.0.0-fat.jar:na]
        at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:418) ~[custom-service-adapter-1.0.0-fat.jar:na]
        at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:440) ~[custom-service-adapter-1.0.0-fat.jar:na]
        at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:873) ~[custom-service-adapter-1.0.0-fat.jar:na]
        at java.lang.Thread.run(Unknown Source) ~[na:1.8.0_111]
Caused by: rx.exceptions.OnErrorThrowable$OnNextValue: OnError while emitting onNext value: io.knotx.knot.service.service.ServiceEntry.class
        at rx.exceptions.OnErrorThrowable.addValueAsLastCause(OnErrorThrowable.java:109) ~[custom-service-adapter-1.0.0-fat.jar:na]
        at rx.internal.operators.OnSubscribeMap$MapSubscriber.onNext(OnSubscribeMap.java:73) [custom-service-adapter-1.0.0-fat.jar:na]
        ... 91 common frames omitted
```

Temporary workaround for it is to define empty `params`:
```json
          "services": [
            {
              "name": "booksdata",
              "address": "knotx.adapter.service.custom",
              "params": {}
            }
          ]
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -3,6 +3,7 @@ All notable changes to Knot.x will be documented in this file.
 
 ## Unreleased
 List of changes that are finished but not yet released in any final version.
+- [PR-290](https://github.com/Cognifide/knotx/pull/290) - allow definig services without default `params` configured
 - [PR-289](https://github.com/Cognifide/knotx/pull/289) - upgraded versions: Vert.x to 3.4.1 and RxJava to 1.2.7
 - [PR-285](https://github.com/Cognifide/knotx/pull/285) - fixed handling of Multiple headers with the same name
 - [PR-278](https://github.com/Cognifide/knotx/pull/278) - fixed closing files in Filesystem Repository

--- a/knotx-example/knotx-example-app/src/main/resources/knotx-example-app.json
+++ b/knotx-example/knotx-example-app/src/main/resources/knotx-example-app.json
@@ -44,8 +44,7 @@
             },
             {
               "name": "labelsRepository",
-              "address": "knotx.adapter.service.http",
-              "params": {}
+              "address": "knotx.adapter.service.http"
             }
           ]
         }

--- a/knotx-example/knotx-example-app/src/test/resources/knotx-test-app.json
+++ b/knotx-example/knotx-example-app/src/test/resources/knotx-test-app.json
@@ -44,8 +44,7 @@
             },
             {
               "name": "labelsRepository",
-              "address": "knotx.adapter.service.http",
-              "params": {}
+              "address": "knotx.adapter.service.http"
             }
           ]
         }

--- a/knotx-knot/knotx-knot-service/src/main/java/io/knotx/knot/service/service/ServiceEntry.java
+++ b/knotx-knot/knotx-knot-service/src/main/java/io/knotx/knot/service/service/ServiceEntry.java
@@ -46,7 +46,9 @@ public class ServiceEntry {
   }
 
   public ServiceEntry mergeParams(JsonObject defaultParams) {
-    this.params = defaultParams.copy().mergeIn(this.params);
+    if (defaultParams != null) {
+      this.params = defaultParams.copy().mergeIn(this.params);
+    }
     return this;
   }
 

--- a/knotx-knot/knotx-knot-service/src/test/java/io/knotx/knot/service/service/ServiceEntryTest.java
+++ b/knotx-knot/knotx-knot-service/src/test/java/io/knotx/knot/service/service/ServiceEntryTest.java
@@ -26,12 +26,18 @@ import org.junit.Test;
 
 public class ServiceEntryTest {
 
-  private ServiceKnotConfiguration config;
+  private ServiceKnotConfiguration configWithDefaultParams;
+
+  private ServiceKnotConfiguration configNoDefaultParams;
 
   @Before
   public void setUp() throws Exception {
-    JsonObject configFile = new JsonObject(FileReader.readText("service-correct.json"));
-    this.config = new ServiceKnotConfiguration(configFile);
+    configWithDefaultParams = new ServiceKnotConfiguration(
+        new JsonObject(FileReader.readText("service-correct.json"))
+    );
+    configNoDefaultParams = new ServiceKnotConfiguration(
+        new JsonObject(FileReader.readText("service-correct-no-params.json"))
+    );
   }
 
 
@@ -40,7 +46,7 @@ public class ServiceEntryTest {
     ServiceEntry serviceEntry = new ServiceEntry(
         new Attribute("data-knotx-service-first", "first-service"),
         new Attribute("data-knotx-params-first", "{\"path\":\"first-service\"}"));
-    serviceEntry.mergeParams(config.getServices().stream().findFirst().get().getParams());
+    serviceEntry.mergeParams(configWithDefaultParams.getServices().stream().findFirst().get().getParams());
     Assert.assertEquals("first-service", serviceEntry.getParams().getString("path"));
   }
 
@@ -49,7 +55,7 @@ public class ServiceEntryTest {
     ServiceEntry serviceEntry = new ServiceEntry(
         new Attribute("data-knotx-service-first", "first-service"),
         new Attribute("data-knotx-params-first", "{}"));
-    serviceEntry.mergeParams(config.getServices().stream().findFirst().get().getParams());
+    serviceEntry.mergeParams(configWithDefaultParams.getServices().stream().findFirst().get().getParams());
     Assert.assertEquals("/service/mock/first.json", serviceEntry.getParams().getString("path"));
   }
 
@@ -58,9 +64,17 @@ public class ServiceEntryTest {
     ServiceEntry serviceEntry = new ServiceEntry(
         new Attribute("data-knotx-service-first", "first-service"),
         new Attribute("data-knotx-params-first", "{\"name\":\"first-service-name\"}"));
-    serviceEntry.mergeParams(config.getServices().stream().findFirst().get().getParams());
+    serviceEntry.mergeParams(configWithDefaultParams.getServices().stream().findFirst().get().getParams());
     Assert.assertEquals("/service/mock/first.json", serviceEntry.getParams().getString("path"));
     Assert.assertEquals("first-service-name", serviceEntry.getParams().getString("name"));
   }
 
+  @Test
+  public void mergePayload_whenNoDefaultParams_expectDefinedParamsUsed() throws Exception {
+    ServiceEntry serviceEntry = new ServiceEntry(
+        new Attribute("data-knotx-service-first", "first-service"),
+        new Attribute("data-knotx-params-first", "{\"path\":\"some-other-service.json\"}"));
+    serviceEntry.mergeParams(configNoDefaultParams.getServices().stream().findFirst().get().getParams());
+    Assert.assertEquals("some-other-service.json", serviceEntry.getParams().getString("path"));
+  }
 }

--- a/knotx-knot/knotx-knot-service/src/test/resources/service-correct-no-params.json
+++ b/knotx-knot/knotx-knot-service/src/test/resources/service-correct-no-params.json
@@ -1,0 +1,9 @@
+{
+  "services": [
+    {
+      "name" : "first-service",
+      "address" : "knotx.core-adapter",
+      "cacheKey": "first"
+    }
+  ]
+}

--- a/knotx-performance-tests/config-knotx/author-config.json
+++ b/knotx-performance-tests/config-knotx/author-config.json
@@ -145,8 +145,7 @@
             },
             {
               "name": "labelsRepository",
-              "address": "knotx.adapter.service.http",
-              "params": {}
+              "address": "knotx.adapter.service.http"
             }
           ]
         }

--- a/knotx-performance-tests/config-knotx/publish-config.json
+++ b/knotx-performance-tests/config-knotx/publish-config.json
@@ -146,8 +146,7 @@
             },
             {
               "name": "labelsRepository",
-              "address": "knotx.adapter.service.http",
-              "params": {}
+              "address": "knotx.adapter.service.http"
             }
           ]
         }

